numba/cuda/tests/cudapy/test_sync.py

```diff
@@ -71,6 +90,30 @@ def test_dyn_shared_memory(self):
         compiled[1, shape, 0, ary.size * 4](ary)
         self.assertTrue(np.all(ary == 2 * np.arange(ary.size, dtype=np.int32)))
 
+    def test_threadfence_codegen(self):
+        compiled = cuda.jit("void(int32[:])")(use_threadfence)
+        ary = np.zeros(10, dtype=np.int32)
+        compiled[1, 1](ary)
```

Comment:
Would that work? A thread fence doesn't actually synchronize threads, it only ensures writes are consistent (mainly because of the weakly ordered memory model, if I understand correctly). Perhaps at the warp level all threads are guaranteed to run in lockstep? I don't know enough about the CUDA execution model.